src/AbstractPNLPModels.jl

```diff
@@ -0,0 +1,286 @@
+module ModAbstractPSNLPModels
+
+using ExpressionTreeForge
+using LinearOperators
+using ADNLPModels, NLPModels, NLPModelsJuMP
+using Printf, Statistics, LinearAlgebra
+
+import Base.show
+
+export AbstractPartiallySeparableNLPModel, AbstractPQNNLPModel, SupportedNLPModel
+export ElementFunction
+export update_nlp, hess_approx, update_nlp!
+
+abstract type AbstractPartiallySeparableNLPModel{T, S} <: AbstractNLPModel{T, S} end
+abstract type AbstractPQNNLPModel{T,S} <: AbstractPartiallySeparableNLPModel{T, S} end
+
+""" Accumulate the supported NLPModels. """
+SupportedNLPModel = Union{ADNLPModel, MathOptNLPModel}
+
+"""
+    ElementFunction
+
+A type that gathers the information indentifying an element function in a `PartiallySeparableNLPModel`, and its properties.
+`ElementFunction` has fields:
+
+* `i`: the index of the element function;
+* `index_element_tree`: the index occupied in the element-function vector after the deletion of redundant element functions;
+* `variable_indices`: list of elemental variables of `ElementFunction`;
+* `type`: `constant`, `linear`, `quadratic`, `cubic` or `general`;
+* `convexity_status`: `constant`, `linear`, `convex`, `concave` or `unknown`.
+"""
+mutable struct ElementFunction
+  i::Int # the index of the function 1 ≤ i ≤ N
+  index_element_tree::Int # 1 ≤ index_element_tree ≤ M
+  variable_indices::Vector{Int} # ≈ Uᵢᴱ
+  type::ExpressionTreeForge.Type_calculus_tree
+  convexity_status::ExpressionTreeForge.M_implementation_convexity_type.Convexity_wrapper
+end
+
+include("common_methods.jl")
+
+# NLPModels interface for AbstractPartiallySeparableNLPModel
+"""
+    f = obj(nlp, x)
+
+Evaluate `f(x)`, the objective function of `nlp` at `x`.
+"""
+function NLPModels.obj(
+  nlp::P,
+  x::AbstractVector{T},
+) where {T, S, P <: AbstractPartiallySeparableNLPModel{T, S}} 
```

Comment:
```suggestion
  nlp::AbstractPartiallySeparableNLPModel{T, S},
  x::AbstractVector{T},
) where {T, S} 
```
You can make this simplification everywhere.